_posts/2023-02-19-SOCKSification-Outreachy-Internship.md

```diff
@@ -0,0 +1,31 @@
+---
+
+SOCKSification refers to the process of rerouting network traffic from an application via a SOCKS proxy server. This method is frequently used for security, anonymity, and circumventing network constraints. There are numerous techniques for SOCKSification, but in this article, we will focus on two main approaches that were considered.
+
+The first method includes intercepting the application's connect and send system calls. When an application wants to connect to a remote server, it performs a connect system call with the target IP address and port specified. Following that, the send system call is used to transfer data through the established connection.
+
+With this method, we intercept the connect syscall and change the target IP address and port to those of the SOCKS proxy server. Then we intercept the send syscall which will contain a pointer (in tracee memory space) to the buffer that the application wants to send. We can prepend a SOCKS handshake to that buffer.  While this method can accomplish SOCKSification, it necessitates changing the tracee's memory, which is not memory safe and might cause security problems.
+
+The second method, which is safer, includes utilizing the pidfd getfd capability, which was added to Linux v5.6, to duplicate the file descriptor of the established connection. When an application connects, the connect syscall returns a file descriptor, which may be used to transmit and receive data through the connection. We can replicate the file descriptor from the tracee to the tracer using the pidfd getfd functionality without having to touch the tracee's memory.
+
+To SOCKSify the program, we intercept the connect syscall and change the target IP address and port to those of the SOCKS proxy server. The exit of the connect syscall, which returns the file descriptor of the established socket, is then intercepted. Using pidfd getfd, we copy this file descriptor to the tracer and execute a SOCKS5 handshake with the tracer's socket. After the handshake, we resume the tracee, and all future network traffic gets SOCKSified.
+
+SOCKSification occurs in the function [Socksify.](https://github.com/robertmin1/heteronculous-horklump/blob/test-3/main.go#L412) This function allows an application running on a Linux system to use the SOCKS5 protocol to route network traffic through a proxy server.
+
+The function begins by checking whether the "one circuit" configuration option is set. If not, it initializes a set of authentication data and generates a random index to select a username and password pair to use for authentication with the proxy server. All connections done will have different authentication details, this is important in order to prevent vulnerability to a Sybil attack, applications that connect to a P2P network (such as the Bitcoin network) may need to avoid using the same Tor circuit for all connections. This is because a malicious Tor exit relay can then block the application's view of the network.
```

Comment:
Maybe elaborate a bit here, explaining that we generate 10 SOCKS user+pass pairs and assign each connection to a randomly selected slot from those 10? Otherwise it may sound like an application that opens 1000 TCP connections will create 1000 Tor circuits, which would be a DoS issue.

Also maybe mention that this is also helpful for things like download managers, since it avoids slowdowns caused by an unlucky slow Tor circuit.